engine/src/main/java/org/pentaho/di/trans/steps/excelinput/staxpoi/StaxPoiSheet.java

```diff
@@ -0,0 +1,444 @@
+      cellsArray.add( null );
+    }
+    // add the found Cell
+    cellsArray.add( cell );
+  }
+
+  @Override
+  public String getName() {
+    return sheetName;
+  }
+
+  @Override
+  public int getRows() {
+    return numRows;
+  }
+
+  @Override
+  public KCell getCell( int colnr, int rownr ) {
+    if ( rownr == 0 && colnr < headerRow.size() ) {
+      // only possible to return header
+      return new StaxPoiCell( headerRow.get( colnr ), rownr );
+    }
+    // if random access this will be very expensive
+    KCell[] row = getRow( rownr );
+    if ( row != null && rownr < row.length ) {
+      return row[colnr];
+    }
+    return null;
+  }
+
+  private KCellType getCellType( String cellType, String cellStyle, boolean isFormula ) {
+    // numeric type can be implicit or 'n'
+    if ( cellType == null || cellType.equals( "n" ) ) {
+      // the only difference between date and numeric is the cell format
+      if ( isDateCell( cellStyle ) ) {
+        return isFormula ? KCellType.DATE_FORMULA : KCellType.DATE;
+      }
+      return isFormula ? KCellType.NUMBER_FORMULA : KCellType.NUMBER;
+    }
+    switch ( cellType ) {
+      case "s":
+        return KCellType.LABEL;
+      case "b":
+        return isFormula ? KCellType.BOOLEAN_FORMULA : KCellType.BOOLEAN;
+      case "e":
+        // error
+        return KCellType.EMPTY;
+      case "str":
+      default:
+        return KCellType.STRING_FORMULA;
+    }
+  }
+
+  @VisibleForTesting
+  protected boolean isDateCell( String cellStyle ) {
+    if ( cellStyle != null ) {
+      int styleIdx = Integer.parseInt( cellStyle );
+      CTXf cellXf = styles.getCellXfAt( styleIdx );
+      if ( cellXf != null ) {
+        // need id for builtin types, format if custom
+        short formatId = (short) cellXf.getNumFmtId();
+        String format = styles.getNumberFormatAt( formatId );
+        return DateUtil.isADateFormat( formatId, format );
+      }
+    }
+    return false;
+  }
+
+  private Object parseValue( KCellType type, String vContent ) {
+    if ( vContent == null ) {
+      return null;
+    }
+    try {
+      switch ( type ) {
+        case NUMBER:
+        case NUMBER_FORMULA:
+          return Double.parseDouble( vContent );
+        case BOOLEAN:
+        case BOOLEAN_FORMULA:
+          return vContent.equals( "1" );
+        case DATE:
+        case DATE_FORMULA:
+          Double xlDate = Double.parseDouble( vContent );
+          return DateUtil.getJavaDate( xlDate, DATE_TZ );
+        case LABEL:
+        case STRING_FORMULA:
+        case EMPTY:
+        default:
+          return vContent;
+      }
+    } catch ( Exception e ) {
+      return vContent;
+    }
+  }
+
+  private void resetSheetReader() throws IOException, XMLStreamException, InvalidFormatException {
+    sheetReader.close();
+    sheetStream.close();
+    sheetStream = xssfReader.getSheet( sheetId );
+    XMLInputFactory factory = XMLParserFactoryProducer.createSecureXMLInputFactory();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Parsing XML input using a weakly configured XML parser can lead to XML external entity (XXE) injection. To protect against XXE attacks, be sure to securely configure your XML parser.

[Learn more](https://skb.highcastle.a2z.com/implementations/189)